tests/conftest.py

```diff
@@ -875,6 +886,116 @@ def sim_run_engine():
     return RunEngineSimulator()
 
 
+class CallbackSim:
```

Comment:
I think this is really useful. I have a couple of comments:

* Not sure if it's really a simulator as it's more of a convenience collector to capture events from the RunEngine and then assert on them
* I think the matchers are very nice and I wonder if they can be modified so they also work not just with the real RunEngine but also with the `RunEngineSimulator` - there you wouldn't be matching against the documents themselves but against the `Msg`s that will contain the documents.